Add port forward

kazk opened the pull request:
Super rough sketch I mentioned almost 2 months ago. https://github.com/clux/kube-rs/issues/127#issuecomment-754399865

I don't like this at all, but an example with Nginx (from [Python client example](https://github.com/kubernetes-client/python/blob/2febc697157fa57bb9a5bd5ab2f7b9fa4d13c151/examples/pod_portforward.py#L97-L117)) works and can be a starting point for discussion. Any feedback/ideas are appreciated, but I honestly don't remember much, so I might not be able to answer "why?".

#### Protocol

WebSocket messages are binary (`[channel, ...data]`) like `/attach` and `/exec`.

For `/portforward`, channels are associated with the specified ports and each port uses 2 channels: a duplex data channel (`2*i`) and a read-only error channel (`2*i + 1`) for `i`th port.

The first message for a channel is always 3 bytes and describes this association:

1. channel (u8)
2. port number (u16 LE)

#### Idea

EDIT: Maybe this can be done on top of the lower level API provided by kube in this PR. I won't be trying this for now.

Make `Portforwarder` [`tower::MakeService`](https://docs.rs/tower/0.4.6/tower/trait.MakeService.html) (`Service` of `Service`)? I haven't tried, but it might be possible, and it'll make it a lot more usable.

```rust
let portforwarder = pods.portforward("example", &[80]).await?;
// takes port and returns Service<Request<Body>>
let mut forwarder = portforwarder.into_service();
let mut svc = forwarder.call(port).await?;
let response = svc.call(request).await?;
```

---

- [x] Error handling
- [x] Add example similar to `kubectl port-forward example local_port:remote_port`.

clux commented on kube/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+                                txr.send(Message::ToPod(vec)).await.expect("send message")
+                            }
+                        }
+                        Err(err) => panic!("{}", err),
```
what are the error cases here?

kazk replied:
This might be unreachable. It should never happen because reading from duplex never errors. Writing to a closed pipe is an error, but reading is not. https://tokio-rs.github.io/tokio/doc/src/tokio/io/util/mem.rs.html#162-186

clux commented on kube/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+            let mut read_stream = tokio_util::io::ReaderStream::new(r);
+            let mut txr = tx.clone();
+            tokio::spawn(async move {
+                while let Some(res) = read_stream.next().await {
```
interface almost feels like a generalised attach, except we have a task per port.

clux commented on kube/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+                }
+
+                // Odd channels are for errors for (n - 1)/2 th port
+                let is_error = ch % 2 == 1;
+                let port_index = ch / 2;
+                if !initialized[ch] {
+                    if bin.len() != 3 {
+                        panic!("Unexpected initial channel frame size");
+                    }
+
+                    let port = u16::from_le_bytes([bin[1], bin[2]]);
+                    if port != ports[port_index] {
+                        panic!(
+                            "Unexpected port number in initial frame ({} != {})",
+                            port, ports[port_index]
+                        );
+                    }
```
hah, this is hairy

clux commented on kube/src/api/subresource.rs:
```diff
@@ -575,3 +576,73 @@ where
+    K: Clone + DeserializeOwned + Portforwardable,
+{
+    /// Forward ports of a pod
+    pub async fn portforward(&self, name: &str, ports: &[u16]) -> Result<Portforwarder> {
+        let req = self.resource.portforward(name, ports)?;
+        let stream = self.client.connect(req).await?;
+        Ok(Portforwarder::new(stream, ports))
+    }
```
interface still clean :-)

kazk replied:
Yeah, I like this part :laughing:

clux commented:
Hey, this is great! A lot of the hairy channel management (spawning a task for each port makes sense) looks like it's done. Error handling perhaps needs a bit more improvements, but you've already got a nice stream interfaces on top of `Api`. The returned `Portforwarder` letting you drill into the `Port` with the various methods is very user friendly.

Well, insofar as letting people write raw http requests into a stream can be user friendly.
I do think we should strive towards a way to have people dump more ergonomic requests in rather than raw-text, even though I guess they can construct http requests themselves and serialize them manually.

The Service idea does sound interesting though! At least if we can something like the interface you posted.

clux commented:
Ah, forgot about this. Yeah, the example does present the basic functionality, and we can merge it as it stands as an initial version of it. But also feel like it might be better to have the canonical `kubectl port-forward` style example (where we actually listen on that port and forward the tcp traffic) working before releasing.

Then on the other hand, people needing that probably aren't writing up against `kube` do do that atm, and just use `kubectl`.

partiallyordered commented:
I've started using this branch and it's working admirably, thanks both :grinning: . I don't know if it fits here, but consider merging my PR here containing some example code putting hyper on top of the port-forward: https://github.com/kazk/kube-rs/pull/4 before merging #446 .

partiallyordered commented:
Come to think of it, I might also have a tokio-tungstenite-based websocket example in the next few days if that's interesting.

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+struct PortforwarderState {
+    waker: Option<Waker>,
+    finished: bool,
+    ports: Option<Vec<Port>>,
```
does this need to be option wrapped? users have to pass in a slice of ports, so the worst that happens here is that `ports.is_empty()`

kazk replied:
`ports` is never empty because it's required to be non empty in `Request::portforward`.

The reason why `ports` is an `Option` is so we can move it out (see `PortforwarderState::ports()`).

clux replied:
it leads to guaranteed unwraps on the user side though (example line 39 and 40). don't you think we should at least do the unwrap for them if we know these always exists?

clux replied:
> The reason why ports is an Option is so we can move it out (see Portforwarder::ports()).

i thought maybe we could return `&mut [Port]` via `self.ports.as_mut_slice()` and rely on rust ownership to ensure that there's only one user of each port.

kazk replied:
Yeah, `&mut [Port]` sounds good. I actually don't remember why I wanted to allow moving `ports` out of `Portforwarder`. It shouldn't be necessary.

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+    }
+}
+
+pub struct Port {
+    // Data pipe.
+    stream: Option<DuplexStream>,
+    // Error channel.
+    error: Option<ErrorReceiver>,
+}
+
+impl Port {
+    pub(crate) fn new(stream: DuplexStream, error: ErrorReceiver) -> Self {
+        Port {
+            stream: Some(stream),
+            error: Some(error),
+        }
+    }
```
`Port` can only be constructed with an explicit `DuplexStream` and `ErrorReceiver`, do these need to be option wrapped?

clux replied:
i tried to play around wit
```rust
pub struct Port {
  pub stream: DuplexStream,
  pub error: ErrorReceiver,
}
```
and it leads to issues down the line (in examples) when you need to pass the borrowed stream to `tokio_util::io::ReaderStream` - so can see why you'd need this move setup now.

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+
+    /// Data pipe for sending to and receiving from the forwarded port.
+    pub fn stream(&mut self) -> Option<impl AsyncRead + AsyncWrite + Unpin> {
+        self.stream.take()
```
i guess it's because this is also replacing the inner stream with a `None` - but we could do that in another way.
at the very least we could do the unwrap in here (since it's guaranteed)

clux replied:
actually, i guess it's up to the user to not call this more than once - and that's why we probably should not do the unwrap.

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,225 @@
+            let (a, b) = tokio::io::duplex(1024 * 1024);
+            let (tx, rx) = oneshot::channel();
+            ports.push(Port::new(a, rx));
+            errors.push(Some(tx));
```
`errors` vec also seems to also contain `Some` wrapped `tx`'s

clux commented on examples/pod_portforward_hyper_http.rs:
```diff
@@ -0,0 +1,91 @@
+    pods.create(&PostParams::default(), &p).await?;
+
+    // Wait until the pod is running, otherwise we get 500 error.
+    let lp = ListParams::default().fields("metadata.name=example").timeout(10);
+    let mut stream = pods.watch(&lp, "0").await?.boxed();
+    while let Some(status) = stream.try_next().await? {
+        match status {
+            WatchEvent::Added(o) => {
+                info!("Added {}", o.name());
+            }
+            WatchEvent::Modified(o) => {
+                let s = o.status.as_ref().expect("status exists on pod");
+                if s.phase.clone().unwrap_or_default() == "Running" {
+                    break;
+                }
+            }
+            _ => {}
+        }
+    }
```
you could replace this with:

```sh
use kube::runtime::wait::{await_condition, conditions::is_pod_running};
await_condition(pods.clone(), "example", is_pod_running());
```

to shift focus onto the relevant business logic

clux commented on examples/pod_portforward_hyper_http.rs:
```diff
@@ -0,0 +1,91 @@
+    let mut ports = pf.ports().unwrap();
+    let port = ports[0].stream().unwrap();
+
+    let (mut sender, connection) = client::conn::handshake(port).await?;
+
+    // spawn a task to poll the connection and drive the HTTP state
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("Error in connection: {}", e);
+        }
+    });
```
i feel this needs a comment: we already have one task to drive the message machinery in kube-client now, and it might not be clear why we need another task to drive the connection

kazk replied:
I'm not sure what to write. The comment is from [`hyper::client::conn` example](https://docs.rs/hyper/latest/hyper/client/conn/index.html).

This task drives the HTTP state for IO (`impl AsyncRead + AsyncWrite` from `Port`'s `stream()`), so we can send an HTTP request and receive a response instead of writing and reading bytes like the original example. The message loop in `potforward` simply forwards the bytes to the server by prefixing with the associated port, and doesn't know how the bytes were written.

clux replied:
Yeah, it's not a lot of elaborating documentation. Have suggested a minor edit to group the lines that are related to hyper. Not a big deal this though.

clux commented on examples/pod_portforward_hyper_http.rs:
```diff
@@ -0,0 +1,91 @@
+        .body(Body::from(""))
+        .unwrap();
+
+    let (parts, body) = sender.send_request(http_req).await.unwrap().into_parts();
+    assert!(parts.status == 200);
+
+    let body_bytes = body::to_bytes(body).await.unwrap();
+    let body_str = std::str::from_utf8(&body_bytes).unwrap();
+    assert!(body_str.contains("Welcome to nginx!"));
```
key logic - this looks a usable interface - and great example :-)
agree that we can try any potential tower wrapping later. maybe we can open an exploratory issue after.

oh, also: this is using `anyhow::Error` so it should work with less unwraps i think.

clux reviewed (commented):
thanks for fixing this up - we haven'd had real feedback on this in a year, so it's probably fine to add it to get more users. i am happy to approve, just left a few minor nits here first that i think could be easy wins if they work

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,223 @@
+        }
+    }
+
+    /// Data pipe for sending to and receiving from the forwarded port.
```
```suggestion
    /// Data pipe for sending to and receiving from the forwarded port.
    ///
    /// This returns a `Some` on the first call, then a `None` on every subsequent call
```

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,223 @@
+        self.stream.take()
+    }
+
+    /// Future that resolves with any error message or when the error sender is dropped.
+    /// When this resolves, the port should be considered no longer usable.
```
```suggestion
    /// Future that resolves with any error message or when the error sender is dropped.
    ///
    /// This returns a `Some` on the first call, then a `None` on every subsequent call
    /// When the future resolves, the port should be considered no longer usable.
```

clux commented:
I thought the port stuff could be done a little cleaner, but ultimately don't see a good way - so resolved my gripes with some comments. Added some minor suggested tweaks to docs

ultimately though, the functionality here works well and we should probably merge this to invite more users. How are you feeling about it now @kazk?

clux commented:
Leaving this out of the release as it is still a draft PR - although I would be happy merging it.

kazk commented:
Sorry for the delay.
I've refactored to avoid panics, and split each loop into smaller functions. It should be easier to understand and iterate on. It's still missing an example that binds to local port, but we can do it later/ask for help.

kazk commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,335 @@
+    S: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+{
+    let mut writers = Vec::new();
+    // Loops to run concurrently.
+    // We can spawn tasks to run `to_pod_loop` in parallel and flatten the errors, but the other 2 loops
+    // are over a single WebSocket connection and cannot process each port in parallel.
+    let mut loops = Vec::with_capacity(ports.len() + 2);
```
Changed not to spawn tasks because other loops can't run in parallel, and it's easier to manage.

clux replied:
Yeah, it's a lot nicer.

kazk commented:
I got an example to bind local port working. I'll push it after cleaning it up.

clux commented on kube-client/src/api/portforward.rs:
```diff
@@ -0,0 +1,335 @@
+    // are over a single WebSocket connection and cannot process each port in parallel.
+    let mut loops = Vec::with_capacity(ports.len() + 2);
+    // Channel to communicate with the main loop
+    let (sender, receiver) = mpsc::channel::<Message>(1);
```
AFAIKT this gives us one guaranteed send slot. Not sure if this leaves it possible for us to overwrites the last message before sending our end. I don't think so; we read one-by-one from both sides, and send one-by-one so hopefully tokio flushes each message out correctly - but not sure how these things truly propagate below.

kazk replied:
Maybe you're thinking of `tokio::sync::mpsc::channel`?

This is `futures::channel::mpsc::channel` and the [documentation](https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html) says:

> The channel’s capacity is equal to buffer + num-senders. In other words, each sender gets a guaranteed slot in the channel capacity, and on top of that there are buffer “first come, first serve” slots available to all senders.

clux replied:
The channel is shared between ports though, so while each port is guaranteed to be able to have one slot in there.

..but looking into it, it does not look like it's a problem because each `.send` is awaited. For some reason I thought it was just dropped into a ring buffer inside the channel.

clux commented on examples/pod_portforward_hyper_http.rs:
```diff
@@ -0,0 +1,75 @@
+    let ports = pf.ports();
+    let port = ports[0].stream().unwrap();
+
+    let (mut sender, connection) = client::conn::handshake(port).await?;
+
+    // spawn a task to poll the connection and drive the HTTP state
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("Error in connection: {}", e);
+        }
+    });
```
```suggestion
    // let hyper drive the HTTP state in our DuplexStream via a task
    let (mut sender, connection) = hyper::client::conn::handshake(port).await?;
    tokio::spawn(async move {
        if let Err(e) = connection.await {
            eprintln!("Error in connection: {}", e);
        }
    });
```

clux reviewed (approved):
Ah, didn't realise you had such a large refactoring left in you. Thanks a lot, it's definitely cleaner!
Minor comments left, but no real suggestions other than the superficial example doc thing.

clux commented on examples/pod_portforward_bind.rs:
```diff
@@ -0,0 +1,113 @@
+    });
+
+    // Shared `SendRequest<Body>` to relay the request.
+    let context = Arc::new(Mutex::new(sender));
+    let make_service = make_service_fn(move |_conn| {
+        let context = context.clone();
+        let service = service_fn(move |req| handle(context.clone(), req));
+        async move { Ok::<_, Infallible>(service) }
+    });
```
cool quick use of tower!

clux reviewed (approved):
very cool full `port-forward` example in just over 100 lines with lot of helpful technical parts in it!

everything seems like it's working well! don't have any other outstanding comments.
looking forward to have this in :-)

kazk commented:
Yeah, the local port example was easier than I had expected. I think this is good enough as a first version.
Merging now. Not squashing to get @partiallyordered in the [contributors list](https://github.com/kube-rs/kube-rs/graphs/contributors).

clux commented:
Very nice! I'll do a release later on today. [Tons of stuff ready](https://github.com/kube-rs/kube-rs/milestone/7?closed=1).
Will try to put a paragraph in the CHANGELOG about it as it's one of the biggest additions (and technically completes the penultimate requirement for [client-gold](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/csi-new-client-library-procedure.md))
